test/server/cards/02_SHD/units/ReyKeepingThePast.spec.ts

```diff
@@ -0,0 +1,98 @@
+describe('Rey, Keeping the Past', function() {
+    integration(function(contextRef) {
+        describe('Rey\'s Ability', function() {
+            it('ignores Heroism aspect penalty when unit Kylo is controlled', function () {
+                contextRef.setupTest({
+                    phase: 'action',
+                    player1: {
+                        hand: ['rey#keeping-the-past'],
+                        groundArena: ['kylo-ren#killing-the-past', 'pyke-sentinel'],
+                        base: 'dagobah-swamp',
+                        leader: 'darth-vader#dark-lord-of-the-sith'
+                    }
+                });
+
+                const { context } = contextRef;
+
+                context.player1.clickCard(context.rey);
+
+                // Rey should cost 5 since it ignores the villainy aspect
+                expect(context.player1.countExhaustedResources()).toBe(5);
+            });
+
+            it('heals 2 and gives Shield token to a non-heroism unit', function () {
+                contextRef.setupTest({
+                    phase: 'action',
+                    player1: {
+                        groundArena: ['rey#keeping-the-past', 'pyke-sentinel']
+                    }
+                });
+
+                const { context } = contextRef;
+
+                context.setDamage(context.pykeSentinel, 2);
+                context.player1.clickCard(context.rey);
+                expect(context.player1).toBeAbleToSelectExactly([context.rey, context.pykeSentinel]);
+                context.player1.clickCard(context.pykeSentinel);
+                expect(context.pykeSentinel.damage).toBe(0);
+                expect(context.pykeSentinel).toHaveExactUpgradeNames(['shield']);
+            });
+
+            it('heals 2 and gives Shield token to an enemy non-heroism non-villany unit', function () {
```

Comment:
Do you mind consolidating these tests into one long one? We try to do that to keep the number lower b/c Jasmine tests are very slow and we have finite github actions time in the free tier (though it's bad unit test practice in general sadly). Can see Rukh tests for kind of an example of how we stack multiple on-attack tests into one, we set up one big describe with all the important pieces.

Similar note for the new Kylo test above

Reply:
Refactored